scheduler: top utility

fwbrasil opened the pull request:
I'm working on https://github.com/getkyo/kyo/issues/347 and need more visibility on the execution of the scheduler to debug. This PR introduces a `top` utility that can be used in two ways:

- **Console**: It's possible to print the information to the console periodically by setting `-Dscheduler.enableTopConsoleMs=<<interval>>`. Disabled by default.
- **JMX**: The functionality is exposed as an mbean and can be accessed the main class `kyo.scheduler.util.Top`. Enabled by default since there shouldn't be a significant performance penalty.

The utility is lightweight and designed to be used in production as well. Demo:

[![asciicast](https://asciinema.org/a/FFZUkfAuMy5cnb16Q0N8KSKBR.svg)](https://asciinema.org/a/FFZUkfAuMy5cnb16Q0N8KSKBR)

fwbrasil commented on kyo-scheduler/jvm/src/main/scala/kyo/scheduler/Scheduler.scala:
```diff
@@ -212,10 +217,53 @@ final class Scheduler(
         )
     }
+
+    def status(): Scheduler.Status = {
```
The scheduler classes now expose `status` methods, which can be useful outside of the `top` utility as well.

fwbrasil commented on kyo-scheduler/jvm/src/main/scala/kyo/scheduler/Scheduler.scala:
```diff
@@ -212,10 +217,53 @@ final class Scheduler(
+        admission: Admission.AdmissionStatus,
+        concurrency: Concurrency.AdmissionStatus
+    ) {
+        infix def -(other: Status): Status =
```
Logic to reset counters on each observation.

fwbrasil commented on kyo-scheduler/jvm/src/main/scala/kyo/scheduler/regulator/Admission.scala:
```diff
@@ -18,13 +20,8 @@ final class Admission(
-        }
-    }
+    private val rejected = new LongAdder
+    private val allowed  = new LongAdder
```
I've moved some metrics to fields to inspect them. `UnsafeCounter` doesn't offer an API to obtain the current value since it can be a no-op impl.

fwbrasil commented on kyo-bench/src/main/scala/kyo/bench/Bench.scala:
```diff
@@ -12,7 +12,11 @@ import zio.UIO
+        "-Dcom.sun.management.jmxremote",
+        "-Dcom.sun.management.jmxremote.port=1099",
+        "-Dcom.sun.management.jmxremote.authenticate=false",
+        "-Dcom.sun.management.jmxremote.ssl=false"
```
to make it easier to debug benchmarks

hearnadam commented on kyo-scheduler/jvm/src/main/scala/kyo/scheduler/util/Top.scala:
```diff
@@ -0,0 +1,141 @@
+
+        // Worker table rows
+        status.workers.foreach { w =>
+            if (w != null) {
```
`ne null`

hearnadam commented on kyo-scheduler/jvm/src/main/scala/kyo/scheduler/util/Top.scala:
```diff
@@ -0,0 +1,141 @@
+    private def print(status: Scheduler.Status): String = {
+        val sb = new StringBuilder()
+
+        // Header
+        sb.append(f"Kyo Scheduler Status      | LoadAvg: ${status.loadAvg}%1.4f   | Flushes: ${status.flushes}\n\n")
+        sb.append("==============================================================================================\n")
+        // Worker table header
+        sb.append("Worker | Running | Blocked | Stalled | Load  | Exec  | Preempt | Done  | Stolen | Lost | Thread \n")
+        sb.append("================================================================================================\n")
```
minor: these could be multi-line strings, right?

hearnadam reviewed (approved):
Awesome! This will be incredibly useful!

fwbrasil commented:
I've addressed the feedback and made a few rendering improvements:

![image](https://github.com/getkyo/kyo/assets/831175/18a825dd-4096-4d27-9112-a3b86304c1d5)
